src/components/menu/menu.svelte

```diff
@@ -0,0 +1,235 @@
+          // Note: This should line up with Reacts key type, but we don't want
+          // to depend on React in this layer, so we just define it manually.
+          key?: string | number | null
+
+          // value?: string
+          children?: any
+        }
+      }
+    }
+  }
+</script>
+
+<script lang="ts">
+  import clickOutside from '../../svelteDirectives/clickOutside'
+
+  let isOpen = false
+  let popup: HTMLDivElement
+  let button: HTMLButtonElement
+
+  const slots = $$slots
+
+  $: menuItems = Array.from(
+    (
+      popup?.querySelector('.leo-menu-popup slot') as HTMLSlotElement
+    )?.assignedElements() ??
+      popup?.querySelectorAll('.leo-menu-popup > *') ??
+      []
+  )
+
+  $: {
+    for (const [option, index] of menuItems.map((o, i) => [o, i] as const)) {
+      option.setAttribute('tabindex', (index + 1).toString())
+      option.setAttribute('role', 'menu-item')
+    }
+  }
+
+  function selectMenuItem(e: Event) {
+    // Find the option which was clicked on, if any.
+    const item = menuItems.find((item) => e.composedPath().includes(item))
+
+    // If the event was triggered for something which isn't an option don't fire
+    // a change event.
+    if (!item) return
+
+    // Close the popup
+    isOpen = false
+
+    // focus the button again for accessibility - this makes it easy to quickly
+    // change the selection.
+    button.focus()
+  }
+
+  /**
+   * Handles changing the currently focused menu element with the up/down arrow.
+   * @param e The KeyboardEvent
+   */
+  function changeSelection(e: KeyboardEvent) {
+    if (!isOpen || !popup) return
+
+    // Handle closing keys
+    if (e.code === 'Escape') {
+      isOpen = false
+      return
+    }
+
+    // We allow the user to select options with ArrowUp/ArrowDown as well as
+    // tab/shift+tab.
+    let dir = 0
+    if (e.code == 'ArrowUp') dir -= 1
+    if (e.code === 'ArrowDown') dir += 1
+    if (dir === 0) return
+
+    // First, find the currently focusedIndex. If no option is selected, we'll
+    // select the first option. Otherwise, we select the next/previous item (and
+    // wrap around).
+    let focusedIndex = menuItems.findIndex((e) => e.matches(':focus-within'))
+    if (focusedIndex === -1) {
+      focusedIndex = 0
+    } else {
+      focusedIndex += dir
+      if (focusedIndex < 0) focusedIndex = menuItems.length - 1
+      if (focusedIndex >= menuItems.length) focusedIndex = 0
+    }
+
+    ;(menuItems[focusedIndex] as any)?.focus()
+    e.preventDefault() // preventDefault, so we don't accidentally scroll
+  }
+
+  function handleAnchorClick(e) {
+    isOpen = !isOpen
+  }
+
+  function handleBlur(e) {
+    e.stopPropagation()
+    e.preventDefault()
+    isOpen = false
+  }
+</script>
+
+<div class="leo-menu">
+  <button
+    id="anchor-button"
+    class="click-target"
+    aria-haspopup="true"
+    aria-controls="menu"
+    tabindex="0"
+    bind:this={button}
+    on:click|stopPropagation={handleAnchorClick}
+  >
```

Comment:
I guess we just need to make sure that the consumer doesn't provide their own button and only provides non-buttony content

Reply:
I think even if they provide their own button this will be fine, as it has `all: unset`, so the styling should be fine :) Might be a bit weird from an accessibility perspective though